lib/swaps/Swaps.ts

```diff
@@ -141,218 +142,284 @@ class Swaps extends EventEmitter {
         case 'BTC':
-          cmdLnd =  this.lndLtcClient;
           break;
         case 'LTC':
           cmdLnd = this.lndBtcClient;
           break;
-        default:
-          // TODO: handle unsupported currency
-          return;
       }
 
       const request = new lndrpc.SendRequest();
-      request.setAmt(deal.takerAmount!);
-      request.setDestString(deal.takerPubKey);
+      request.setAmt(deal.takerAmount);
+      request.setDestString(deal.takerPubKey!);
       request.setPaymentHashString(deal.r_hash);
 
-      // TODO: use timeout on call
-
       try {
-        const sendPaymentResponse = await cmdLnd.sendPaymentSync(request);
-        if (sendPaymentResponse.getPaymentError()) {
-          throw sendPaymentResponse.getPaymentError();
+        this.setDealStatus(deal, SwapDealStatus.AmountSent);
+        const response = await cmdLnd.sendPaymentSync(request);
+        if (response.getPaymentError()) {
+          this.logger.error('Got error from sendPaymentSync: ' + response.getPaymentError() + ' ' + JSON.stringify(request.toObject()));
+          this.setDealState(deal, SwapDealState.Error, response.getPaymentError());
+          return response.getPaymentError();
         }
 
-        // swap succeeded!
-        const r_preimage = Buffer.from(sendPaymentResponse.getPaymentPreimage_asB64(), 'base64').toString('hex');
-        this.logger.debug(`swap succeeded! ${r_preimage}`);
-        this.emit('swap.completed', deal);
-        responseBody = {
-          r_preimage,
-          dealId,
-        };
+        deal.r_preimage = Buffer.from(response.getPaymentPreimage_asB64(), 'base64').toString('hex');
+        this.setDealStatus(deal, SwapDealStatus.AmountReceived);
+        return deal.r_preimage;
       } catch (err) {
-        this.logger.error(`Got exception from sendPaymentSync ${JSON.stringify(request.toObject())}`, err);
-        this.emit('swap.failed', deal);
+        this.logger.error('Got exception from sendPaymentSync: ' + ' ' + JSON.stringify(request.toObject()) + err.message);
+        this.setDealState(deal, SwapDealState.Error, err.message);
+        return 'Got exception from sendPaymentSync' + err.message;
       }
     } else {
-      // this is not a valid deal
-      this.logger.warn(`received swap request for unrecognized deal ${dealId}`);
-      // TODO: respond with a failure message, possibly penalize peer
+      // If we are here we are the taker
+      this.logger.debug('Executing taker code');
+      this.setDealStatus(deal, SwapDealStatus.AmountReceived);
+      return deal.r_preimage;
     }
 
-    const responsePacket = new packets.SwapResponsePacket(responseBody, requestPacket.header.id);
+  }
 
-    this.logger.debug('Sending swap response to peer: ' + JSON.stringify(responseBody));
+  private setDealState = (deal: SwapDeal, newState: SwapDealState, newStateReason: string): void => {
+    deal.state = newState;
+    deal.stateReason = newStateReason;
+    this.emit('swap.failed', deal);
```

Comment:
Do we want to emit `swap.failed` every time the deal state is updated? If `state` is only used to track the error status, maybe we just rename the field to `errored` and this method to `setError`?

Reply:
Currently, there is only an Error state. So we emit every time (actually once) this is set.

In the future, we may have other states and once we add these we will also update this code. 

I see no reason to change it.

Reply:
What other states might we have? Unless there's something we'd need to handle in the foreseeable future, I'd prefer that we deal with it when/if a new "state" is actually needed. Since this is only being used to indicate whether a swap has errored or not, I think a boolean `errored` field and `errorReason` (instead of `stateReason`, which likewise is only being used to indicate why a swap failed) makes more sense. If there's another state to handle then it's a different story, but I'm not thinking of anything besides "had an error" and "haven't had an error" that's needed for our purposes.

About the `emit` firing every time, my main concern there is that if we were to have other states and someone called `setDealState()` they could unintentionally emit an error event. It also just doesn't follow from the method name. But really my preference would just be to rename this to something like `setError` - at which point emitting `swap.failed` every time makes perfect sense.

Reply:
Changed the code to make sure that `swap.failed` emitted only when setting the state to `Error` (which can happend only once).

We actually have 3 states for the swapdeal. Active, Error, and Completed. With these we can expose a public method named `isFinalState` which can be used to check if all activities ended related to the `swap` or if we are expecting more `events`.

we already discussed about `removing objects` like `swapdeal` and `order`. This should not be done. These can be removed from memory if backed to the storage knowing that you can always fetch them back if needed. As such you need to have a mechanism to idenfify active swaps (and orders) from completed/errored ones. The state is for that.

This is relevant to `subscribedOrders` RPC call and buy/sell with wait option. Will add a comment there (#123) shortly.

Reply:
Also didn't see until after reviewing the latest changes, I made a comment about this below, over redundnancy between status completed and state completed.

As for removing swap deals, I agree w/ removing from memory but persisting to the database. For orders we are deliberately not persisting them to the database *at all*, that should be a separate discussion from swaps if you think that should change though.

Reply:
Maybe. Maybe today. Maybe even not today.
But if we add something between the switch and the assignment,  that is
blocking for some reason this may break.  When this happened to trace it
would be a nightmare.
So, the code may look nicer with one switch but clearly not robust.

I don't see swap different from orders. swap is a mechanism which is part of order execution. IMHO both must be persistent.